crates/app-host/src/event_log_fetcher.rs

```diff
@@ -0,0 +1,210 @@
+    fn from(value: EthError) -> Self {
+        Self::Eth(value)
+    }
+}
+
+pub struct EventLogFetcher {
+    event_log_parser: EventLogParser,
+    l1_client: L1Client,
+    scroll_chain_address: Address,
+    max_size_per_fetch: u64,
+    fetched_block_number: u64,
+    finalized_block_number: u64,
+
+    commit_batch_tx: Sender<CommitBatchEvent>,
+    finalize_batch_tx: Sender<FinalizeBatchEvent>,
+}
+
+impl EventLogFetcher {
+    pub fn new(
+        commit_batch_tx: Sender<CommitBatchEvent>,
+        finalize_batch_tx: Sender<FinalizeBatchEvent>,) -> Self {
+        todo!()
+    }
+
+    async fn get_latest_finalized_block(&self) -> Result<u64, EventLogError> {
+        let block = self.l1_client.get_block_by_number(BlockNumberOrTag::Finalized).await.map_err(EthError::from)?;
+        if block.is_none() {
+            return Err(EventLogError::Fetcher("get empty block".into()));
+        }
+        match block.unwrap().header.number {
+            Some(n) => Ok(n),
+            None => Err(EventLogError::Fetcher("no block number in header".into()))
+        }
+    }
+
+    async fn fetch_logs(&self) -> Result<(), EventLogError> {
+        let last_finalize_block = self.get_latest_finalized_block().await?;
+        let event_signatures = vec![
+            ScrollChain::CommitBatch::SIGNATURE_HASH,
+            ScrollChain::FinalizeBatch::SIGNATURE_HASH,
+        ];
+
+        let from: u64 = self.fetched_block_number;
+        let mut to = from + self.max_size_per_fetch;
+        if to > last_finalize_block {
+            to = last_finalize_block;
+        }
+
+        let logs = self.l1_client.get_logs(self.scroll_chain_address, event_signatures, from, to).await.map_err(EthError::from)?;
+
+        for log in logs {
+            match log.topic0() {
+                Some(&ScrollChain::CommitBatch::SIGNATURE_HASH) => {
+                    match self.event_log_parser.parse_commit_batch_log(log).await {
+                        Ok(event) => {
+                            self.commit_batch_tx.send(event).await;
+                        },
+                        Err(err) => {
+                            todo!()
+                        }
+                    }
+                },
+                Some(&ScrollChain::FinalizeBatch::SIGNATURE_HASH) => {
+                    match self.event_log_parser.parse_finalize_batch_log(log).await {
+                        Ok(event) => {
+                            self.finalize_batch_tx.send(event).await;
+                        },
+                        Err(err) => {
+                            todo!()
+                        }
+                    }
+                },
+                _ => {
+                    todo!()
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn start(&self) -> () {
+        let mut interval = interval(Duration::from_secs(10));
+        loop {
+            interval.tick().await;
+
+            self.fetch_logs();
```

Comment:
We should not wait 10s during initial catch up

Reply:
resolved